# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(x):
    return x + 2

print(add_two(3))

5


In [3]:
# lambda param: one-line function
lambda x: x + 2

<function __main__.<lambda>(x)>

In [4]:
# IIFE - Immediately Invoked Function Expression
# Execute a lambda function immediately after defining the function
# Syntax: (lambda param: expression)(arg)
(lambda x: x + 2)(3)

5

#### Saving to a Variable

In [5]:
f = lambda x,y: x + y + 2
print(f)

<function <lambda> at 0x000001FAC95B8940>


In [6]:
f(3,2)

7

#### Multiple Inputs

In [7]:
# Syntax: lambda param1, param2, etc: expression
(lambda x,y,z: x * y * z)(3,5,8)

120

In [ ]:
# As an IIFE - (lambda param1, param2, etc: expression)(arg1, arg2, etc.)


In [9]:
def apply_func(func, arg):
    return func(arg)

In [10]:
def my_math(x):
    return x*x


In [14]:
apply_func(my_math, 5)

25

#### Passing a Function into a Function

In [17]:
apply_func(lambda x: x*x, 5)


25

#### Returning a Function from a Function

In [ ]:
def create_multiplier(num):
    def

#### Returning a Lambda from a Function

In [18]:
def new_multiplier(num):
    return lambda x: x * num


four = new_multiplier(4)
print(four)

<function new_multiplier.<locals>.<lambda> at 0x000001FACB0F9090>


In [19]:
four(10)

40

#### If Statements within Lambdas

In [20]:
def double_pos_square_negs(number):
    if number >= 0:
        return number * 2
    else:
        return number ** 2
    
    
print(double_pos_square_negs(10))
print(double_pos_square_negs(20))
print(double_pos_square_negs(-10))
print(double_pos_square_negs(-20))

20
40
100
400


In [ ]:
# Ternary: value_if_true if condition else value_if_false


In [23]:
def do_something(num):
    return num * 2 if num > 10 else num + 2
do_something(5)

7

In [24]:
(lambda num: num *2 if num > 10 else num +2) (5)

7

In [26]:
#write a lambda function that takes a name and age and returns 
# a string that says this [name] is [this many] years old 
(lambda name, age: f"{name} is {age} years old!")('Mason', 22)

' Mason is 22 years old!'

In [25]:
#write a lambda function that takes 3 numbers and returns those numbers added together 
(lambda x, y, z: x + y + z)(1,2,3)

6

In [34]:
help(sorted)
#Creates a NEW list sorted

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [33]:
help(list.sort)
#DOES NOT create a new list

Help on method_descriptor:

sort(self, /, *, key=None, reverse=False)
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



In [29]:
from random import randint

my_class = [
    {'Name': 'Valencia', "IQ": randint(1,249)},
    {'Name': 'Mason', "IQ": randint(1,249)},
    {'Name': 'Cynthia', "IQ": randint(1,249)},
    {'Name': 'Sheridan', "IQ": randint(1,249)}
]

print(my_class)

[{'Name': 'Valencia', 'IQ': 48}, {'Name': 'Mason', 'IQ': 185}, {'Name': 'Cynthia', 'IQ': 89}, {'Name': 'Sheridan', 'IQ': 106}]


In [30]:
sorted(my_class, key=lambda student: student['IQ'])

[{'Name': 'Valencia', 'IQ': 48},
 {'Name': 'Cynthia', 'IQ': 89},
 {'Name': 'Sheridan', 'IQ': 106},
 {'Name': 'Mason', 'IQ': 185}]

In [32]:
sorted(my_class, key=lambda student: student['IQ'], reverse=True)

[{'Name': 'Mason', 'IQ': 185},
 {'Name': 'Sheridan', 'IQ': 106},
 {'Name': 'Cynthia', 'IQ': 89},
 {'Name': 'Valencia', 'IQ': 48}]

In [35]:
print(my_class)

[{'Name': 'Valencia', 'IQ': 48}, {'Name': 'Mason', 'IQ': 185}, {'Name': 'Cynthia', 'IQ': 89}, {'Name': 'Sheridan', 'IQ': 106}]


In [38]:
my_class.sort(key=lambda student: student['IQ'], reverse=True)
print(my_class)

[{'Name': 'Mason', 'IQ': 185}, {'Name': 'Sheridan', 'IQ': 106}, {'Name': 'Cynthia', 'IQ': 89}, {'Name': 'Valencia', 'IQ': 48}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

In [39]:
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [45]:
def squared(num1, num2):
    if num1 < 10 and num2 < 10:
        return (num1**2, num2**2)
    else:
        return (num1, num2)
    
print(squared(11,11))
print(squared(8,8))
print(squared(8,12))


nums1 = [1,10,3,1,6]
nums2 = [4,1,20,3,15]

new_list = []
for i in range(len(nums1)):
    new_list.append(squared(nums1[i], nums2[i]))
print(new_list)
  

(11, 11)
(64, 64)
(8, 12)
[(1, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


In [47]:
# map(function, *iterables -> List, Tuple, String, Set, Dictionary, etc)
squared_nums_map = map(squared, nums1, nums2)
print(list(squared_nums_map))

[(1, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


#### Using Lambda's with Map

In [54]:
squared_nums_map = map((lambda num1, num2: (num1**2, num2**2) if num1 <10 and num2 <10 else (num1,num2)), nums1, nums2)
print(list(squared_nums_map))

[(1, 16), (10, 1), (3, 20), (1, 9), (6, 15)]


#### In-Class Exercise #1 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [60]:
nums = [4,2,6,8,15]
# [7, 3, 11, 15, 29]

double_minus_1 = map((lambda num: num*2 -1), nums)
print(list(double_minus_1))

[7, 3, 11, 15, 29]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

In [61]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



#### Syntax

In [62]:
bool('Hello')

True

In [65]:
# Syntax: filter(function_to_apply, iterable -> List,Tuple,Set,Dict,String,etc)
list(filter(None, [True, True, False, "", "", "string", [], [1,2,3,4,5]]))

[True, True, 'string', [1, 2, 3, 4, 5]]

#### Using Lambda's with Filter()

In [71]:
names = ["Dainel", "Leisa", "Sheridan", "Desiree", "Keanu", "Santa", "Danny Devito"]
def d_names(name):
    if name[0].lower() == 'd':
        return True
    else:
        return False

the_d_names = filter(d_names, names)
print(list(the_d_names))

['Dainel', 'Desiree', 'Danny Devito']


In [75]:
def d_names(name):
    True if name[0].lower() == 'd' else False

lamb_ds = filter(lambda name: True if name[0].lower() == 'd' else False, names)
list(lamb_ds)

['Dainel', 'Desiree', 'Danny Devito']

#### In-Class Exercise #2 <br>
<p>Using the filter function, create a list out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [84]:
from statistics import mean

nums = [2,7, 4.2, 1.6, 9, 4.4 ,4.9] #4.73
    
below = filter(lambda num: True if num <= mean(nums) else False, nums)

list(below)

[2, 4.2, 1.6, 4.4]

In [85]:
below = filter(lambda num: num <= mean(nums), nums)

list(below) #more compact 

[2, 4.2, 1.6, 4.4]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [86]:
from functools import reduce

help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, iterable[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence
    or iterable, from left to right, so as to reduce the iterable to a single
    value.  For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the iterable in the calculation, and serves as a default when the
    iterable is empty.



In [87]:
list_1 = [2,4,6,8,10]

def subtract_nums(num1, num2):
    return num1 - num2
results = reduce(subtract_nums, list_1)
print(results)

-26


#### Using Lambda's with Reduce()

In [92]:
the_sum = reduce((lambda aggregator, current_value: aggregator + current_value) , list_1)
print(the_sum)

30


#### In-Class Exercise #3 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [94]:
multi = reduce((lambda aggregator, current_value: aggregator * current_value), list_1)
print(multi)

3840


In [100]:
divide = reduce((lambda aggregator, current_value: aggregator / current_value), list_1)
print(divide)

0.0010416666666666667


In [103]:
letters = ["C", "o", "m", "e", " ", "o", "n", " ", "D", "o","w","n"]
full_phrase = reduce((lambda phrase, letter: phrase + letter), letters)
print(full_phrase)

Come on Down


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [108]:
def recursive(i):
    if i == 4: #STOP LOOPING ----- Base Case (Case we need to stop)
        return
    print('hi!')
    i += 1
    recursive(i)

In [194]:
recursive(0)

hi!
hi!
hi!
hi!


#### Implementing a Base Case

#### Writing a Factorial Function

In [113]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# ...
# 1! = 1

def factorial(num):
    if num <= 1:
        print("factorial(1)=1")
        return 1
    else:
        print(f"Factorial({num}={num}) * factorial({num -1})")
        return num * factorial(num - 1)
    
factorial(5)

Factorial(5=5) * factorial(4)
Factorial(4=4) * factorial(3)
Factorial(3=3) * factorial(2)
Factorial(2=2) * factorial(1)
factorial(1)=1


120

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [141]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York", "DC"]
  
cities = filter((lambda city: city.strip() and city[0].isalpha()), places)
# cities2 = filter(lambda city: True if city.strip() and city[1].lower() else False, places)

print(list(cities))
# print(list(cities2))

['Argentina', 'San Diego', 'Boston', 'New York', 'DC']
['Argentina', 'San Diego', 'Boston', 'New York', 'DC']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [174]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

last_name_sort = (lambda names: names.split() and names[-1]), (author)

print(list(last_name_sort))

[<function <lambda> at 0x000001FACB32BD90>, ['Joel Carter', 'Victor aNisimov', 'Andrew P. Garfield', 'David hassELHOFF', 'Gary A.J. Bernstein']]


In [146]:
"a">"B" #remeber this ;) 

True

### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [191]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]

In [192]:
# def celsius(cities):
#     f = []
#     for city in cities:
#         temp = (9/5) * city[-1] + 32
#         f.append(temp)
#     return f
        
# print(celsius(places))

c_to_f = map(lambda places: (places[0], (9/5) * places[-1] + 32), places)
print(list(c_to_f))

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>
<code>fib(1) = 0
fib(2) = 1
fib(3) = 1
fib(4) = 2
fib(5) = 3
fib(6) = 5
fib(7) = 8
fib(8) = 13
fib(9) = 21
fib(10) = 34
fib(11) = 55
fib(12) = 89
fib(13) = 144
</code>

In [1]:
def recursive(i):
    if i == 4: #STOP LOOPING ----- Base Case (Case we need to stop)
        return
    print('hi!')
    i += 1
    recursive(i)

In [8]:
# fib(1) = 0
# fib(2) = 1
# fib(3) = 1
# fib(4) = 2
# fib(5) = 3
# fib(6) = 5
# fib(7) = 8
# fib(8) = 13
# fib(9) = 21
# fib(10) = 34
# fib(11) = 55
# fib(12) = 89
# fib(13) = 144

In [1]:
# def fib(num):
#     if num == 14:
#         return
#     prev_num = 0
#     sum = num + prev_num
#     print(f"fib{num} = {sum}")
#     prev_num += 1
#     fib(num)
def fib(num):
    if num == 14:
        return
    prev_num = 0
    sum = num + prev_num
    num += 1
    print(f"fib{num} = {sum}")
    fib(num) 

In [ ]:
fib(0)

### Exercise #5<br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [ ]:
def my_gen(number):
    pass

for num in my_gen(10):
    print(num)
    
# 100
# 81
# 64
# 49
# 36
# 25
# 16
# 9
# 4
# 2
# 1
# 0